server/src/test/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorTests.java

```diff
@@ -152,6 +596,129 @@ public void testUnmappedMissingGeoPoint() throws IOException {
         });
     }
 
+    public void testSingleValuedStringHashed() throws IOException {
+        final CardinalityAggregationBuilder aggregationBuilder = new CardinalityAggregationBuilder("name").field("str_value.hash");
+        final MappedFieldType mappedFieldTypes = new KeywordFieldMapper.KeywordFieldType("str_value.hash");
+
+        testAggregation(aggregationBuilder, new MatchAllDocsQuery(), iw -> {
+            iw.addDocument(
+                singleton(new SortedDocValuesField("str_value.hash", new BytesRef(hash("one".getBytes(StandardCharsets.UTF_8)))))
+            );
+            iw.addDocument(
+                singleton(new SortedDocValuesField("unrelatedField.hash", new BytesRef(hash("two".getBytes(StandardCharsets.UTF_8)))))
+            );
+            iw.addDocument(
+                singleton(new SortedDocValuesField("str_value.hash", new BytesRef(hash("three".getBytes(StandardCharsets.UTF_8)))))
+            );
+            iw.addDocument(
+                singleton(new SortedDocValuesField("str_value.hash", new BytesRef(hash("four".getBytes(StandardCharsets.UTF_8)))))
+            );
+        }, card -> {
+            assertEquals(3, card.getValue(), 0);
+            assertTrue(AggregationInspectionHelper.hasValue(card));
+        }, mappedFieldTypes);
+    }
+
+    public void testAsSubAggregation() throws IOException {
+        final MappedFieldType mappedFieldTypes[] = {
+            new KeywordFieldMapper.KeywordFieldType("str_value"),
+            new NumberFieldMapper.NumberFieldType("number", NumberFieldMapper.NumberType.LONG) };
+
+        final AggregationBuilder aggregationBuilder = new TermsAggregationBuilder("terms").field("str_value")
+            .missing("unknown")
+            .subAggregation(AggregationBuilders.cardinality("cardinality").field("number"));
+
+        final Directory directory = newDirectory();
+        final RandomIndexWriter indexWriter = new RandomIndexWriter(random(), directory);
+        final int numDocs = 10;
```

Comment:
One of the `testCase` style methods in the superclass ought to be able to do this stuff for you. It's nice to use them mostly because they add extra randomization to the reduction and the circuit breaker.